### PR TITLE
refactor: Refactor launcher_utils.py

### PR DESCRIPTION
**Pull request by prmukherj**

### CODE REVIEW

File: src/ansys/fluent/core/launcher/custom_exceptions.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+class GPUSolverSupportError(ValueError):
+    """Raised when an unsupported Fluent version is specified."""
+
+    def __init__(self):
+        super().__init__("Fluent GPU Solver is only supported for 3D.")
+
+
+class IpPortNotProvided(ValueError):
+    """Raised when ip and port are not specified."""
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/custom_exceptions.py:22]:
```suggestion
    """Raised when IP address and port are not specified."""
```

File: src/ansys/fluent/core/launcher/custom_exceptions.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+
+class UnexpectedKeywordArgument(TypeError):
+    """Raised when a valid keyword argument is not specified."""
+
+    pass
+
+
+class DockerContainerLaunchNotSupported(SystemError):
+    """Raised when docker container launch is not supported."""
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/custom_exceptions.py:35]:
```suggestion
    """Raised when Docker container launch is not supported."""
```

File: src/ansys/fluent/core/launcher/custom_exceptions.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        ("-g" in additional_arg_list) or ("-gu" in additional_arg_list)
+    ):
+        raise InvalidArgument("Unsupported '-g' and '-gu' on windows platform.")
+
+
+def _raise_non_gui_exception_in_windows(
+    ui: FluentUI, product_version: FluentVersion
+) -> None:
+    """Fluent user interface mode lower than ``FluentUI.HIDDEN_GUI`` is not supported in
+    Windows in Fluent versions lower than 2024 R1."""
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/custom_exceptions.py:64]:
```suggestion
    Windows in Fluent versions earlier than 2024 R1."""
```

**Comment by prmukherj** [comment]
@PipKat, Please ignore this PR. I'll be replaced by another soon. Thank you.

File: src/ansys/fluent/core/launcher/launch_string.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        if argval is None and v.get("fluent_required") is True:
+            argval = default
+        if argval is not None:
+            allowed_values = v.get("allowed_values")
+            if allowed_values and argval not in allowed_values:
+                if default is not None:
+                    old_argval = argval
+                    argval = default
+                    logger.warning(
+                        f"Specified value '{old_argval}' for argument '{k}' is not an allowed value ({allowed_values}), default value '{argval}' is going to be used instead."
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launch_string.py:69]:
```suggestion
                        f"Specified value '{old_argval}' for argument '{k}' is not an allowed value ({allowed_values}). Default value '{argval}' is going to be used instead."
```

File: src/ansys/fluent/core/launcher/launcher_arguments.py [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+from ansys.fluent.core.session_pure_meshing import PureMeshing
+from ansys.fluent.core.session_solver import Solver
+from ansys.fluent.core.session_solver_icing import SolverIcing
+from ansys.fluent.core.utils.fluent_version import FluentVersion
+from ansys.fluent.core.utils.networking import find_remoting_ip
+import ansys.platform.instancemanagement as pypim
+
+
+class LaunchMode(Enum):
+    """An enumeration over supported Fluent launch modes."""
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_arguments.py:29]: Our standard is to start class descriptions with a verb ending is "s" or "es". I haven't commented on this earlier, but this is the standard that most PyAnsys libraries are following.

File: src/ansys/fluent/core/launcher/launcher_arguments.py [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+    """An enumeration over supported Fluent modes."""
+
+    MESHING_MODE = (Meshing, "meshing")
+    PURE_MESHING_MODE = (PureMeshing, "pure-meshing")
+    SOLVER = (Solver, "solver")
+    SOLVER_ICING = (SolverIcing, "solver-icing")
+
+    @staticmethod
+    def get_mode(mode: str) -> "FluentMode":
+        """Returns the FluentMode based on the provided mode string.
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_arguments.py:47]:
```suggestion
        """Get the Fluent mode based on the provided mode string.
```
Docstrings should use plain English whenever possible.

File: src/ansys/fluent/core/launcher/launcher_arguments.py [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+
+        Parameters
+        ----------
+        mode : str
+            mode
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_arguments.py:52]:
```suggestion
            Mode.
```

File: src/ansys/fluent/core/launcher/launcher_arguments.py [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+
+        Returns
+        -------
+        FluentMode
+            Fluent mode
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_arguments.py:57]:
```suggestion
            Fluent mode.
```

File: src/ansys/fluent/core/launcher/launcher_arguments.py [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+        allowed_modes = []
+        for m in FluentMode:
+            allowed_modes.append(m.value[1])
+            if mode == m.value[1]:
+                return m
+        raise DisallowedValuesError("mode", mode, allowed_modes)
+
+    @staticmethod
+    def is_meshing(mode: "FluentMode") -> bool:
+        """Returns whether the current mode is meshing.
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_arguments.py:73]:
```suggestion
        """Check if the current mode is meshing.
```

File: src/ansys/fluent/core/launcher/launcher_arguments.py [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+def get_fluent_exe_path(**launch_argvals) -> Path:
+    """Get Fluent executable path. The path is searched in the following order.
+
+    1. ``product_version`` parameter passed with ``launch_fluent``.
+    2. The latest ANSYS version from ``AWP_ROOTnnn``` environment variables.
+
+    Returns
+    -------
+    Path
+        Fluent executable path
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_arguments.py:97]:
```suggestion
    """Get the path for the Fluent executable file. The search for the path is performed in this order.

    1. Look at the ``product_version`` parameter passed with the ``launch_fluent()`` method.
    2. Look at the latest Ansys version in ``AWP_ROOTnnn``` environment variables.

    Returns
    -------
    Path
        Fluent executable path
```
Ansys is always lowercase except for in trademark statements.

File: src/ansys/fluent/core/launcher/launcher_arguments.py [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+    product_version = launch_argvals.get("product_version")
+    if product_version:
+        return get_exe_path(get_fluent_root(FluentVersion(product_version)))
+
+    # 2. the latest ANSYS version from AWP_ROOT environment variables
+    return get_exe_path(get_fluent_root(FluentVersion.get_latest_installed()))
+
+
+def _is_windows():
+    """Check if the current operating system is windows."""
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_arguments.py:126]:
```suggestion
    """Check if the current operating system is Windows."""
```

File: src/ansys/fluent/core/launcher/launcher_arguments.py [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+    return platform.system() == "Windows"
+
+
+def _get_mode(mode: Optional[Union[FluentMode, str, None]] = None):
+    """Updates the session information."""
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_arguments.py:131]:
```suggestion
    """Update the session information."""
```

File: src/ansys/fluent/core/launcher/launcher_arguments.py [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+        mode = FluentMode.get_mode(mode)
+
+    return mode
+
+
+def _get_running_session_mode(
+    fluent_connection: FluentConnection, mode: Optional[FluentMode] = None
+):
+    """Get the mode of the running session if the mode has not been mentioned
+    explicitly."""
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_arguments.py:145]:
```suggestion
    """Get the mode of the running session if the mode has not been
    explicitly given."""
```

File: src/ansys/fluent/core/launcher/launcher_arguments.py [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+                if fluent_connection.scheme_eval.scheme_eval("(cx-solver-mode?)")
+                else "meshing"
+            )
+        except Exception as ex:
+            raise InvalidPassword() from ex
+    return session_mode.value[0]
+
+
+def _get_fluent_launch_mode(start_container, container_dict, scheduler_options):
+    """Get Fluent launch mode.
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_arguments.py:161]:
```suggestion
    """Get the Fluent launch mode.
```

File: src/ansys/fluent/core/launcher/launcher_arguments.py [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+    ip: Optional[str] = None,
+    port: Optional[int] = None,
+    password: Optional[str] = None,
+):
+    """Get server connection information of an already running session."""
+    if not (ip and port) and not server_info_file_name:
+        raise IpPortNotProvided()
+    if (ip or port) and server_info_file_name:
+        logger.warning(
+            "The ip and port will be extracted from the server-info file and their explicitly specified values will be ignored."
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_arguments.py:215]:
```suggestion
            "The IP address and port are extracted from the server-info file and their explicitly specified values are ignored."
```
Use present tense.

File: src/ansys/fluent/core/launcher/launcher_arguments.py [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+    return kwargs
+
+
+def _process_kwargs(kwargs):
+    """Verify whether keyword arguments are valid or not.
+
+    Parameters
+    ----------
+    kwargs: Any
+        Provided keyword arguments.
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_arguments.py:261]:
```suggestion
        Keyword arguments.
```

File: src/ansys/fluent/core/launcher/launcher_arguments.py [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+
+
+def _process_invalid_args(dry_run, fluent_launch_mode, argvals):
+    """Get invalid arguments.
+
+    Parameters
+    ----------
+    dry_run: bool
+        If dry running a container start,
+        ``launch_fluent()`` will return the configured ``container_dict``.
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_arguments.py:286]:
```suggestion
        Whether to dry runa container start. If ``True``, the
        ``launch_fluent()`` method returns the configured ``container_dict``.
```

File: src/ansys/fluent/core/launcher/launcher_arguments.py [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+            )
+
+
+def _get_standalone_launch_fluent_version(
+    product_version: Union[FluentVersion, str, None]
+) -> Optional[FluentVersion]:
+    """Determine the Fluent version during the execution of the ``launch_fluent()``
+    method in standalone mode.
+
+    The search for the version is performed in this order.
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_arguments.py:324]:
```suggestion
    The search for the version is performed in this order:
```

File: src/ansys/fluent/core/launcher/launcher_utils.py [hunk 19]
```diff
@@ -2,85 +2,19 @@
...
-    """An enumeration over supported Fluent launch modes."""
-
-    STANDALONE = 1
-    PIM = 2
-    CONTAINER = 3
-    SLURM = 4
-
-
 def check_docker_support():
     """Checks whether Python Docker SDK is supported by the current system."""
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/launcher_utils.py:19]:
```suggestion
    """Check whether Python Docker SDK is supported by the current system."""
```

File: src/ansys/fluent/core/launcher/pim_launcher.py [hunk 20]
```diff
@@ -205,3 +212,85 @@ def __call__(self):
...
+    Parameters
+    ----------
+    session_cls: Union[type(Meshing), type(PureMeshing), type(Solver), type(SolverIcing)]
+        Session type.
+    start_transcript: bool
+        Whether to start streaming the Fluent transcript in the client. The
+        default is ``True``. You can stop and start the streaming of the
+        Fluent transcript subsequently via method calls on the session object.
+    product_version : str, optional
+        Select an installed version of ANSYS. The string must be in a format like
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/pim_launcher.py:243]:
```suggestion
        Ansys version to use. The string must be in a format like
```

File: src/ansys/fluent/core/launcher/pim_launcher.py [hunk 21]
```diff
@@ -205,3 +212,85 @@ def __call__(self):
...
+        ``"23.2.0"`` (for 2023 R2) matching the documented version format in the
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/pim_launcher.py:244]:
```suggestion
        ``"23.2.0"`` (for 2023 R2), matching the documented version format in the
```

File: src/ansys/fluent/core/launcher/pim_launcher.py [hunk 22]
```diff
@@ -205,3 +212,85 @@ def __call__(self):
...
+        Whether to clean up and exit Fluent when Python exits or when garbage
+        is collected for the Fluent Python instance. The default is ``True``.
+    mode : FluentMode, optional
+        Whether to launch Fluent remotely in meshing mode. The default is
+        ``FluentMode.SOLVER``.
+    dimensionality : str, optional
+        Geometric dimensionality of the Fluent simulation. The default is ``None``,
+        in which case ``"3d"`` is used. Options are ``"3d"`` and ``"2d"``.
+    file_transfer_service : optional
+        File transfer service. Uploads/downloads files to/from the server.
```
Comment (PipKat) [inline src/ansys/fluent/core/launcher/pim_launcher.py:257]:
```suggestion
        File transfer service for uploading or downloading files to or from the server.
```

**Comment by PipKat** [comment]
> @PipKat, Please ignore this PR. I'll be replaced by another soon. Thank you.

Sorry! I just say your comment saying to ignore the PR. Oh well--the advise provided in my comments still apply!

**Pull request closed without merging**